logstash-core/lib/logstash/agent.rb

```diff
@@ -75,9 +75,12 @@ def execute
       Stud.interval(@reload_interval) { reload_state! }
     else
       while !Stud.stop?
-        if clean_state? || running_pipelines?
-          sleep 0.5
-        else
+        if running_user_defined_pipelines?
+          sleep(0.5)
+        elsif running_pipelines?
+          logger.debug("Shutting down system pipelines")
+          shutdown_pipelines
+        elsif clean_state? || !running_pipelines?
```

Comment:
we don't need an elsif condition here to do `shutdown_pipelines` since we do an `agent.shutdown` after `agent.execute` terminates (see runner code).